题目来自公众号
https://mp.weixin.qq.com/s/SyC_LLQL8AU3i6wYNlOdNQ

1、一行代码实现1--100之和

In [6]:
sum(range(1,101))

5050

2、如何在一个函数内部修改全局变量

In [7]:

a=5
def fn():
    global a
    a=4
    
fn()
print(a)


4


3、列出5个python标准库

In [8]:
import numpy
import tensorflow
import datetime

C:\Users\liangh\Anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


4、字典如何删除键和合并两个字典

In [20]:
d1={"one":1, "two":2}
d2={"three":3, "four":4}

d1.update(d2)
print(d1)

del d1["two"]


{'one': 1, 'two': 2, 'three': 3, 'four': 4}


5、谈下python的GIL

6、python实现列表去重的方法

In [21]:
a=[1,2,3,3]
a=list(set(a))
print(a)

[1, 2, 3]


7、fun(*args,**kwargs)中的*args,**kwargs什么意思？

In [35]:
def demo(a,b,*args):
    print("a:"+a)
    print("b:"+b)
    for arg in args:
        print("args:",arg)
        
demo('a','b','c','d')        
        
def demo2(a,*args,**kargs):
    print("a", a)
    print("args", args)
    print("kargs", kargs)
    
demo2('a','b','c',d='')



a:a
b:b
args: c
args: d
a a
args ('b', 'c')
kargs {'d': ''}


8、python2和python3的range（100）的区别

In [51]:
from collections import Iterable
from collections import Iterator

print(type(range(100)))
print(isinstance(range(100), Iterable))
print(isinstance(range(100), Iterator))

for x in range(5):
    print(x)
    
i=iter(range(5))
while True:
    try:
        v=next(i)
        print(v)
    except StopIteration:
        break
        

<class 'range'>
True
False
0
1
2
3
4
0
1
2
3
4


9、一句话解释什么样的语言能够用装饰器?

10、python内建数据类型有哪些



In [57]:
print(int())
print(bool())
print(str())
print(list())
print(tuple())
print(dict())

0
False

[]
()
{}


11、简述面向对象中__new__和__init__区别

In [63]:
class MyClass:
    def __init__(self):
        print(id(self))
        print("init")
        
    def __new__(cls):
        print("new")        
        obj=object.__new__(cls)
        print(id(obj))
        return obj
cls=MyClass()


new
82772264
82772264
init


12、简述with方法打开处理文件帮我我们做了什么？

In [70]:
with open('python110.ipynb','rb') as f:
    print(f.read(100))
    


b'{\n "cells": [\n  {\n   "cell_type": "markdown",\n   "metadata": {},\n   "source": [\n    "1\xe3\x80\x81\xe4\xb8\x80\xe8\xa1\x8c\xe4\xbb\xa3\xe7\xa0'


13、列表[1,2,3,4,5],请使用map()函数输出[1,4,9,16,25]，并使用列表推导式提取出大于10的数，最终输出[16,25]

In [88]:
a=[1,2,3,4,5]
b_=map(lambda x: x**2, a)
print(list(b_))

b=map(lambda x: x**2, a)
c=[x for x in b if x > 10]
print(c)

[1, 4, 9, 16, 25]
[16, 25]


14、python中生成随机整数、随机小数、0--1之间小数方法

In [102]:
import random
print(random.random())
print(random.randint(0,10))

import numpy as np
print(np.random.random(5))
print(np.random.randint(5,10,size=(2,2)))


0.08168540925528966
1
[0.78392345 0.87393776 0.09639987 0.56392186 0.14751041]
[[5 9]
 [5 6]]


15、避免转义给字符串加哪个字母表示原始字符串？

In [103]:
print(r"\\abc")

\\abc


16、`<div class="nam">中国</div>`，用正则匹配出标签里面的内容（“中国”），其中class的类名是不确定的

In [107]:
import re
reg='<div class=".*">(.*?)</div>'
a=re.findall(reg,r'6、<div class="nam">中国</div>')
print(a)

['中国']


17、python中断言方法举例

In [109]:
a=3
assert(a>0)


18、数据表student有id,name,score,city字段，其中name中的名字可有重复，需要消除重复行,请写sql语句

?


19、10个Linux常用命令

ls  pwd  cd  touch  rm  mkdir  tree  cp  mv  cat  more  grep  echo 

20、python2和python3区别？列举5个

21、列出python中可变数据类型和不可变数据类型，并简述原理

In [115]:
int
bool
float
str
tuple

tuple

22、s = "ajldjlajfdljfddd"，去重并从小到大排序输出"adfjl"

In [123]:
s = "ajldjlajfdljfddd"
a="".join(sorted(list(set(s))))
print(a)

adfjl


23、用lambda函数实现两个数相乘

In [126]:
f=lambda x,y: x*y
f(3,4)

12

24、字典根据键从小到大排序



In [132]:
dic={"name":"zs","age":18,"city":"深圳","tel":"1362626627"}

print(sorted(list(dic.items()),key=lambda x: x[0]))


[('age', 18), ('city', '深圳'), ('name', 'zs'), ('tel', '1362626627')]


25、利用collections库的Counter方法统计字符串每个单词出现的次数"kjalfj;ldsjafl;hdsllfdhg;lahfbl;hl;ahlf;h"



In [135]:
from collections import Counter
a="kjalfj;ldsjafl;hdsllfdhg;lahfbl;hl;ahlf;h"
ans=Counter(a)
print(ans)

Counter({'l': 9, ';': 6, 'h': 6, 'f': 5, 'a': 4, 'j': 3, 'd': 3, 's': 2, 'k': 1, 'g': 1, 'b': 1})


26、字符串a = "not 404 found 张三 99 深圳"，每个词中间是空格，用正则过滤掉英文和数字，最终输出"张三  深圳"

In [137]:
import re

27、filter方法求出列表所有奇数并构造新列表，a =  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]



In [141]:
a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
list(filter(lambda x: x%2==1,a))


[1, 3, 5, 7, 9]

28、列表推导式求列表所有奇数并构造新列表，a =  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]



In [143]:
a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[x for x in a if x%2 == 1]

[1, 3, 5, 7, 9]

29、正则re.complie作用

30、a=（1，）b=(1)，c=("1") 分别是什么类型的数据？

In [148]:
print(type((1,)))
print(type((1)))
print(type(("1")))


<class 'tuple'>
<class 'int'>
<class 'str'>


31、两个列表[1,5,7,9]和[2,2,6,8]合并为[1,2,2,3,6,7,8,9]

In [152]:
a=[1,5,7,9]
b=[2,2,6,8]
a.extend(b)
a.sort()
print(a)

[1, 2, 2, 5, 6, 7, 8, 9]


32、用python删除文件和用linux命令删除文件方法

python：os.remove(文件名)

linux:       rm  文件名

33、log日志中，我们需要用时间戳记录error,warning等的发生时间，请用datetime模块打印当前时间戳 “2018-04-01 11:38:54”

In [166]:
import datetime
n=datetime.datetime.now()
n.strftime("%Y-%m-%d %H:%M:%S")

'2019-10-11 13:34:17'

34、数据库优化查询方法


35、请列出你会的任意一种统计图（条形图、折线图等）绘制的开源库，第三方也行



36、写一段自定义异常代码



In [175]:
def catchException():
    try:
        for i in range(5):
            print(i)
            if i > 2:
                raise Exception("too large")
    except Exception as ex:
        print(ex)
    
catchException()   

0
1
2
3
too large


37、正则表达式匹配中，（.*）和（.*?）匹配区别？


38、简述Django的orm



39、[[1,2],[3,4],[5,6]]一行代码展开该列表，得出[1,2,3,4,5,6]


In [185]:
a=[[1,2],[3,4],[5,6]]
b=[x for ar in a for x in ar]
print(b)

[1, 2, 3, 4, 5, 6]


40、x="abc",y="def",z=["d","e","f"],分别求出x.join(y)和x.join(z)返回的结果



In [191]:
x="abc"
y="def"
z=["d","e","f"]
print(x.join(y))
print(x.join(z))



dabceabcf
dabceabcf


41、举例说明异常模块中try except else finally的相关意义



In [197]:

try:
    raise Exception("abc")
except Exception:
    print("except")
else:
    print('else')
finally:
    print("finally")


except
finally


42、python中交换两个数值



In [200]:
a,b=1,2
b,a=a,b
print(a,b)

2 1


43、举例说明zip（）函数用法



In [202]:
for x,y in zip(range(1,5), range(100,200)):
    print(x,y)

1 100
2 101
3 102
4 103


44、a="张明 98分"，用re.sub，将98替换为100



45、写5条常用sql语句



46、a="hello"和b="你好"编码成bytes类型



In [204]:
a="hello"
b="你好"


47、[1,2,3]+[4,5,6]的结果是多少？



In [207]:
[1,2,3]+ [4,5,6]

[1, 2, 3, 4, 5, 6]

48、提高python运行效率的方法



49、简述mysql和redis区别



50、遇到bug如何处理

51、正则匹配，匹配日期2018-03-20



52、list=[2,3,5,4,9,6]，从小到大排序，不许用sort，输出[2,3,4,5,6,9]



In [216]:
a=[2,3,5,4,9,6]
out=[]
while len(a) != 0:
    m=min(a)
    out.append(m)
    a.remove(m)
print(out)

[2, 3, 4, 5, 6, 9]


53、写一个单列模式



In [225]:
class S:
    __instance=None
    def __new__(cls,v):
        if cls.__instance is None:
            cls.__instance=object.__new__(cls)
            cls.__instance.v=v
        return cls.__instance

    
print(id(S(1)))
print(id(S(2)))
a=S(1)
b=S(2)
print(a.v)

225305544
225305544
1


54、保留两位小数

55、求三个方法打印结果


56、列出常见的状态码和意义



57、分别从前端、后端、数据库阐述web项目的性能优化



58、使用pop和del删除字典中的"name"字段，dic={"name":"zs","age":18}



In [234]:
dic={"name":"zs","age":18}

dic.pop('name')
del dic['age']


59、列出常见MYSQL数据存储引擎



60、计算代码运行结果，zip函数历史文章已经说了，得出[("a",1),("b",2)，("c",3),("d",4),("e",5)]



61、简述同源策略



62、简述cookie和session的区别



63、简述多线程、多进程



64、简述any()和all()方法



In [239]:
print(any([0,1]))
print(all([0,1]))


True
False


65、IOError、AttributeError、ImportError、IndentationError、IndexError、KeyError、SyntaxError、NameError分别代表什么异常



66、python中copy和deepcopy区别



In [253]:
import copy
a="abc"
print(id(a))
print(id(copy.copy(a)))
print(id(copy.deepcopy(a)))


a=["a","b"]
aa=copy.copy(a)
print(id(a))
print(id(aa))
print(id(a[0]),id(a[1]))
print(id(aa[0]),id(a[1]))

a=[["a","c"],"b"]
aa=copy.copy(a)
print(id(a))
print(id(aa))
print(id(a[0]),id(a[1]))
print(id(aa[0]),id(a[1]))

a=[["a","c"],"b"]
aa=copy.deepcopy(a)
print(id(a))
print(id(aa))
print(id(a[0]),id(a[1]))
print(id(aa[0]),id(a[1]))


31210720
31210720
31210720
81721224
82717000
31129416 31127736
31129416 31127736
81720904
225153416
81778696 31127736
81778696 31127736
82716808
82717000
81721224 31127736
82716616 31127736


67、列出几种魔法方法并简要介绍用途



In [259]:
class A:
    def __str__(a):
        return "A"
        
        
print(A())

A


68、C:\Users\ry-wu.junya\Desktop>python 1.py 22 33命令行启动程序并传参，print(sys.argv)会输出什么数据？



In [262]:
import sys
print(sys.argv)

['C:\\Users\\liangh\\Anaconda3\\lib\\site-packages\\ipykernel_launcher.py', '-f', 'C:\\Users\\liangh\\AppData\\Roaming\\jupyter\\runtime\\kernel-040c8d91-4c05-44a2-818f-d9dccc550b81.json']


69、请将[i for i in range(3)]改成生成器



In [266]:
print(type([i for i in range(3)]))
print((i for i in range(3)))


<class 'list'>
<generator object <genexpr> at 0x000000000D6FD1A8>


70、a = "  hehheh  ",去除收尾空格



In [268]:
a = " hehheh "
a.strip()

'hehheh'

71、举例sort和sorted对列表排序，list=[0,-1,3,-10,5,9]

In [272]:
a=[0,-1,3,-10,5,9]
sorted(a)

[-10, -1, 0, 3, 5, 9]

72、对list排序foo = [-5,8,0,4,9,-4,-20,-2,8,2,-4],使用lambda函数从小到大排序



In [275]:
foo = [-5,8,0,4,9,-4,-20,-2,8,2,-4]
sorted(foo, key=lambda x:x)

[-20, -5, -4, -4, -2, 0, 2, 4, 8, 8, 9]

73、使用lambda函数对list排序foo = [-5,8,0,4,9,-4,-20,-2,8,2,-4]，输出结果为

[0,2,4,8,8,9,-2,-4,-4,-5,-20]，正数从小到大，负数从大到小



In [278]:
foo = [-5,8,0,4,9,-4,-20,-2,8,2,-4]
sorted(foo, key=lambda x: (x<0,abs(x)))

[0, 2, 4, 8, 8, 9, -2, -4, -4, -5, -20]

74、列表嵌套字典的排序，分别根据年龄和姓名排序



In [281]:
foo = [{"name":"zs","age":19},{"name":"ll","age":54},
        {"name":"wa","age":17},{"name":"df","age":23}]
sorted(foo, key=lambda x : (x['name'],x['age']))


[{'name': 'df', 'age': 23},
 {'name': 'll', 'age': 54},
 {'name': 'wa', 'age': 17},
 {'name': 'zs', 'age': 19}]

75、列表嵌套元组，分别按字母和数字排序




76、列表嵌套列表排序，年龄数字相同怎么办？



77、根据键对字典排序（方法一，zip函数）



78、根据键对字典排序（方法二,不用zip)



79、列表推导式、字典推导式、生成器



In [287]:
print(type([i for i in range(3)]))
print(type((i for i in range(3))))
print(type({i:i*i for i in range(3)}))


<class 'list'>
<class 'generator'>
<class 'dict'>


80、最后出一道检验题目，根据字符串长度排序，看排序是否灵活运用



81、举例说明SQL注入和解决办法



82、s="info:xiaoZhang 33 shandong",用正则切分字符串输出['info', 'xiaoZhang', '33', 'shandong']



83、正则匹配以163.com结尾的邮箱



84、递归求和



85、python字典和json字符串相互转化方法



86、MyISAM 与 InnoDB 区别：



87、统计字符串中某字符出现次数



In [289]:
a=["abc","d","abc"]
a.count("abc")

2

88、字符串转化大小写



In [291]:
a="Abc"
print(a.lower())
print(a.upper())

abc
ABC


89、用两种方法去空格



In [293]:
a="a b c"
a.replace(" ","")

'abc'

90、正则匹配不是以4和7结尾的手机号



91、简述python引用计数机制



92、int("1.4"),int(1.4)输出结果？



In [297]:
int(1.4)
int("1.4")

ValueError: invalid literal for int() with base 10: '1.4'

94、正则表达式匹配第一个URL



95、正则匹配中文



96、简述乐观锁和悲观锁



97、r、r+、rb、rb+文件打开模式区别



98、Linux命令重定向 > 和 >>



99、正则表达式匹配出`<html><h1>www.itcast.cn</h1></html>`



100、python传参数是传值还是传址？



101、求两个列表的交集、差集、并集



In [300]:
a={"a","b"}
b={"b","c"}
print(a.union(b))
print(a.intersection(b))
print(a.difference(b))


{'b', 'a', 'c'}
{'b'}
{'a'}


102、生成0-100的随机数



103、lambda匿名函数好处



104、常见的网络传输协议



105、单引号、双引号、三引号用法



106、python垃圾回收机制



107、HTTP请求中get和post区别



108、python中读取Excel文件的方法



109、简述多线程、多进程



110、python正则中search和match

